wallet/src/wallet/mod.rs

```diff
@@ -2377,6 +2385,102 @@ impl Wallet {
+    /// specify the `expiration_height` of the lock that defines the height of the local chain at
+    /// which the outpoint becomes spendable.
+    ///
+    /// You must persist the staged change for the lock status to be persistent. To unlock a
+    /// previously locked outpoint, see [`Wallet::unlock_outpoint`].
+    pub fn lock_outpoint(&mut self, outpoint: OutPoint, expiration_height: Option<u32>) {
+        use alloc::collections::btree_map;
+        let lock_value = true;
+
+        // We only stage a change if the lock status changed. Here that means
+        // checking if the outpoint is not currently locked, or if the expiration
+        // height changed.
+        let is_changed = match self.locked_outpoints.entry(outpoint) {
+            btree_map::Entry::Occupied(mut e) => {
+                let mut is_changed = false;
+                let utxo = e.get_mut();
+                if !utxo.is_locked {
+                    utxo.is_locked = lock_value;
+                    is_changed = true;
+                }
+                if utxo.expiration_height != expiration_height {
+                    utxo.expiration_height = expiration_height;
+                    is_changed = true;
+                }
+                is_changed
+            }
+            btree_map::Entry::Vacant(e) => {
+                e.insert(UtxoLock {
+                    outpoint,
+                    is_locked: lock_value,
+                    expiration_height,
+                });
+                true
+            }
+        };
+
+        if is_changed {
+            let utxo_lock = UtxoLock {
+                outpoint,
+                is_locked: lock_value,
+                expiration_height,
+            };
+            self.stage.merge(ChangeSet {
+                locked_outpoints: [(outpoint, utxo_lock)].into(),
+                ..Default::default()
+            });
+        }
+    }
+
+    /// Unlock the wallet output of the specified `outpoint`.
+    pub fn unlock_outpoint(&mut self, outpoint: OutPoint) {
+        use alloc::collections::btree_map;
+        let lock_value = false;
+
+        match self.locked_outpoints.entry(outpoint) {
+            btree_map::Entry::Vacant(..) => {}
+            btree_map::Entry::Occupied(mut e) => {
+                // If the utxo is currently locked, update the lock value and stage
+                // the change.
+                let utxo = e.get_mut();
+                if utxo.is_locked {
+                    utxo.is_locked = lock_value;
+                    let utxo_lock = UtxoLock {
+                        is_locked: lock_value,
+                        ..*utxo
+                    };
+                    self.stage.merge(ChangeSet {
+                        locked_outpoints: [(outpoint, utxo_lock)].into(),
+                        ..Default::default()
+                    });
+                }
+            }
+        }
+    }
+
```

Comment:
What do you think about an `unlock_all_outpoints` method? Just in case the caller needs to reset their state.